flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/PreciseTimestamp.java

```diff
@@ -0,0 +1,234 @@
+			return false;
+		}
+		PreciseTimestamp that = (PreciseTimestamp) obj;
+		return this.compareTo(that) == 0;
+	}
+
+	@Override
+	public String toString() {
+		// yyyy-MM-dd HH:mm:ss
+		StringBuilder timestampString = new StringBuilder(unixTimestampToString(milliseconds));
+		if (precision > 0 && nanoseconds > 0) {
+			// append fraction part
+			timestampString.append('.').append(nonosSecondsToString(nanoseconds, precision));
+		}
+		return timestampString.toString();
+	}
+
+	public long toLong() {
+		return milliseconds;
+	}
+
+	public static PreciseTimestamp fromLong(long milliseconds, int precision) {
+		Timestamp t = new Timestamp(milliseconds);
+		long millis = isCompact(precision) ?
+			zeroLastNDigits(t.getTime(), 3 - precision) : t.getTime();
+		int nanos = (int) zeroLastNDigits((long) t.getNanos(), 9 - precision);
+		return new PreciseTimestamp(millis, nanos, precision);
+	}
+
+	public Timestamp toTimestamp() {
+		return Timestamp.valueOf(this.toString());
+	}
+
+	public static PreciseTimestamp fromTimestamp(Timestamp t, int precision) {
+		String timestampString = t.toString();
+		long millis = timestampStringToUnixDate(timestampString);
+		int nanos = timestampStringToNanoseconds(timestampString);
+		return new PreciseTimestamp(millis, nanos, precision);
+	}
+
+	public LocalDateTime toLocalDateTime() {
+		long epochSeconds = (milliseconds >= 0 || nanoseconds == 0) ?
+			(milliseconds / 1000) : (milliseconds / 1000 - 1);
+		return LocalDateTime.ofEpochSecond(epochSeconds, nanoseconds, ZoneOffset.UTC);
+	}
+
+	public static PreciseTimestamp fromLocalDateTime(LocalDateTime ldt, int precision) {
+		int nanos = (int) zeroLastNDigits((long) ldt.getNano(), 9 - precision);
+		long millis = unixTimestamp(
+			ldt.getYear(),
+			ldt.getMonthValue(),
+			ldt.getDayOfMonth(),
+			ldt.getHour(),
+			ldt.getMinute(),
+			ldt.getSecond()) + (long) (nanos / 1_000_000);
+		return new PreciseTimestamp(millis, nanos, precision);
+	}
+
+	public byte[] toUnscaledBytes() {
+		byte[] bytes8 = new byte[8];
+		long l = milliseconds;
+		for (int i = 0; i < 8; i++) {
+			bytes8[7 - i] = (byte) l;
+			l >>>= 8;
+		}
+
+		if (isCompact()) {
+			return bytes8;
+		}
+
+		byte[] bytes12 = new byte[12];
+		System.arraycopy(bytes8, 0, bytes12, 0, 8);
+		int n = nanoseconds;
+		for (int i = 0; i < 4; i++) {
+			bytes12[11 - i] = (byte) n;
+			n >>>= 8;
+		}
+
+		return bytes12;
+	}
+
+	public static PreciseTimestamp fromUnscaledBytes(byte[] bytes, int precision) {
+		assert bytes.length == 8 || bytes.length == 12;
+
+		long l = 0;
+		for (int i = 0; i < 8; i++) {
+			l <<= 8;
+			l |= (bytes[i] & (0xff));
+		}
+
+		if (isCompact(precision)) {
+			assert bytes.length == 8;
+			return new PreciseTimestamp(l, 0, precision);
+		}
+
+		assert bytes.length == 12;
+		int n = 0;
+		for (int i = 8; i < 12; i++) {
+			n <<= 8;
+			n |= (bytes[i] & (0xff));
+		}
+		return new PreciseTimestamp(l, n, precision);
+	}
+
+	public PreciseTimestamp copy() {
```

Comment:
`copy` should in serializer, There's no need to put it in this class.

Reply:
The serializer has a copy interface which need this:

	@Override
	public PreciseTimestamp copy(PreciseTimestamp from) {
		return from.copy();
	}

Reply:
Immutable class should not have copy method.
You can just return `from` in `serializer.copy`.